### PR TITLE
[google_maps_flutter] Heatmap platform interface

### PR DESCRIPTION
**Pull request by fergusdixon**
## Description

Adds the platform interface for platforms to implement Heatmap functionality.
Future PRs can add support for [Android](https://developers.google.com/maps/documentation/android-sdk/utility/heatmap), [iOS](https://developers.google.com/maps/documentation/ios-sdk/utility/heatmap) and [Web](https://developers.google.com/maps/documentation/javascript/heatmaplayer)

This is heavily reliant on the work done by @SvenSlijkoord in https://github.com/flutter/plugins/pull/2454

The goal of this is to implement Heatmaps in a more modular and easy to review way, starting with the platform interface and allowing contributors to implement individual platforms in seperate PRs.

## Related Issues

https://github.com/flutter/flutter/issues/33586

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by BhupinderRana** [comment]
when can it ready for use heat maps in flutter ? Please it need urgent

**Comment by stuartmorgan** [comment]
We apologize for the long delay in triaging this PR. We’re in the process of overhauling our PR triage system to respond much more quickly, as well as working through the backlog.

We're in the middle of updating this plugin to null-safety, but once that's complete we'd like to work with you on getting this compatible with master and then reviewed.

**Comment by stuartmorgan** [comment]
@ditman @cyanglaz Would one of you be able to do the first review here since you have the most experience with the maps plugin?

@fergusdixon This will need tests to proceed. Can you elaborate on your comment in the PR description? There are existing tests for all the types, which should demonstrate tests for that aspect, and while the method channel doesn't currently have good coverage, there is an example test there that should show the basic structure of testing a new method channel method.

**Comment by ditman** [comment]
I took a quick look, and my earliest, non-exhaustive review would start by: **please remove the `toJson`/`fromJson` from the types that are shared across all platforms.** Those methods are only needed by the "method_channel" implementation, so let's keep them close to that bit of the code, and not part of the types themselves.

Reasons for this suggestion:

* It's very likely that the Web implementation of this feature will need to pass very different objects to the js-interop layer than what's passed back and forth through the Method Channel.
  * _Look at what we have to do [in the web to parse a `CameraUpdate` object](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart#L359-L417) ([type](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart#L109-L112)). (This is probably the most extreme offender of all, but **YIKES!**)_
* Over-reliance in json-backed objects in other parts of this plugin have made it very "stringly-typed" (instead of type-safe). 
  * We've been trying to minimize this in the recent past ([see this](https://github.com/flutter/flutter/issues/70330)), so let's not add more gas to that fire!

PS: Also question if you need all the operator== overrides and hashing. Are those needed for the heatmap implementation? Are those types going to be compared/sorted?

**Comment by fergusdixon** [comment]
Sure it's been over a year since I looked at this but will pull in latest master and get cracking

**Comment by ditman** [comment]
Thanks @fergusdixon! It does make sense that the code is one year old, sorry, didn't notice the dates! (x_X)

**Comment by fergusdixon** [comment]
@ditman @stuartmorgan ready for another review I think
Have added tests as well as removed some of the unneeded methods.
I refactored the `toJson` methods to a utils class, rather than cluttering the method_channel file with them, or do you think they should live in some sort of utils file within the method_channel dir?

**Review by ditman — COMMENTED** [review]
This is looking great, I just added a few questions about the approach. This is much closer to the current implementation of the plugin, thanks for working on this!!

Quick general question: I see that this doesn't modify the `buildView` method (used to pass an "initial heatmap" that renders immediately. Is this intended, that heatmaps must always be set from a "SetHeatmap" method (or similar) in the "MapController"?

This implementation is cool; it seems fully backwards-compatible, but is it good enough for users?

_(PS: I wish all geometry behaved like this, so we wouldn't need such a complex buildView method!)_

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+      visible: visibleParam ?? visible,
+      zIndex: zIndexParam ?? zIndex,
+    );
+  }
+
+  /// Creates a new [Heatmap] object whose values are the same as this
+  /// instance.
+  Heatmap clone() {
+    return copyWith(
+      pointsParam: List<WeightedLatLng>.of(points),
```
[the remark this answers is not in the document]
Comment (fergusdixon) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart:170]: Yea you're right, will fix and add a test for this

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+/// Paints a heatmap on geographical locations on the map.
+@immutable
+class Heatmap {
+  /// Creates an immutable object representing a heatmap on the map.
+  const Heatmap({
+    required this.heatmapId,
+    required this.gradient,
+    this.points = const <WeightedLatLng>[],
```
[the remark this answers is not in the document]
Comment (fergusdixon) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart:101]: I think a heatmap with no points makes sense - for example if a user has a filter on their points, and the filter ends up excluding all points, should they be responsible for checking that before adding it? As far as I understand the Maps SDK should be clever enough to not incur any performance overhead with an empty set of points, I may be wrong though?

In terms of the LatLng & Weighted LatLng, Android and JS seems to support both however [iOS does not](https://github.com/googlemaps/google-maps-ios-utils/blob/e7bd1c648d3d8b4a7bfcda053d66387ef565fcef/src/Heatmap/GMUHeatmapTileLayer.h#L36). I believe the current implementation with forcing WeightedLatLngs that have a default intensity of `1` should be fine?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    return value == typedOther.value;
+  }
+
+  @override
+  int get hashCode => value.hashCode;
+}
+
+/// A pair of latitude and longitude coordinates, stored as degrees with a given intensity.
+@immutable
+class WeightedLatLng {
```
[the remark this answers is not in the document]
Comment (fergusdixon) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart:42]: [iOS does not](https://github.com/googlemaps/google-maps-ios-utils/blob/e7bd1c648d3d8b4a7bfcda053d66387ef565fcef/src/Heatmap/GMUHeatmapTileLayer.h#L36) support LatLngs - should we force WeightedLatLng or let the iOS implementation handle converting that?

**Comment by fergusdixon** [comment]
@ditman I have made many of your suggested changes, thanks so much for the in-depth and helpful review, I really appreciate it!

I believe the only outstanding thing is a decision on whether we should support `LatLng`s as well as `WeightedLatLng`s in the heatmap constructor, as Android and JS allow this, however iOS does not?

Please do let me know if I left anything out, or if there is further feedback on the changes :)

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,9 @@
+## 2.1.5
+
+* Add an interface to define Heatmap behaviour
+
+[0.06666666666666667, 0.13333333333333333, 0.2, 0.26666666666666666, 0.3333333333333333, 0.4, 0.4666666666666667, 0.5333333333333333, 0.6, 0.6666666666666666, 0.7333333333333333, 0.8, 0.8666666666666667, 0.9333333333333333]
```
Comment (fergusdixon) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/CHANGELOG.md:5]:
```suggestion
```

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart:42]: Nah, let's go with the simple implementation for now then! If WeightedLatLngs are easy to create from LatLng for the user, it's probably fine (that way, it's evident that a heatmap is more than a collection of LatLngs, and that WeightedLatLng Is A Thing!)

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart:122]: Yep, let's go with WeightedLatLngs for everything, it's a good call. It also exposes an API that might be unknown (but useful) to the user, like myself.

**Comment by ditman** [comment]
> I believe the only outstanding thing is a decision on whether we should support LatLngs as well as WeightedLatLngs in the heatmap constructor, as Android and JS allow this, however iOS does not?

I think `WeightedLatLng` is a sensible API, especially if it's easy to create from a LatLng. It makes an available optimization very obvious to the end user of the plugin, and they're easy enough to create. Let's go with only the Weighted version of the type for the heatmap values!

> Also with regards to not modifying buildview - here it is is just added directly to the GoogleMap widget constructor instead

Ah, your main plugin code is a little bit old as well (as expected). In the older version, IIRC, all the geometrical entities would get parsed from the `creationParams` object as needed (or directly passed to the native side). Check how the new `buildView` method looks now:

* https://github.com/flutter/plugins/blob/main/packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart#L297-L311

**I think that to wrap this PR up, you should also update your core plugin's (the changes in `package:google_maps_flutter`), to see if you need further changes on the Interface,** and also to see if we can land this in a backwards-compatible manner (or we need a major version bump).

My suggestion about "making the Heatmaps work only from the Map Controller" was something along this lines (even though it's a little bit different to the current state of the art): Instead of passing the heatmap that we want to render through the `GoogleMap` widget (as it is now), we maybe should make the [`_updateHeatmap` method](https://github.com/flutter/plugins/pull/2454/files?authenticity_token=kNuidmuL2dbfS1-VdIcTIE1yuztKjt57WD_nBymyMeA_l6yOwFE2PyXWjhfa-GAMmSmeXaKNvcudUaQ6UcA-pA&file-filters%5B%5D=.dart#diff-ba511c6cbe73561bc93977a3d469e0b2e652682840aa96b2c3c2c443d58d2814R155-R166) public, so heatmaps could work like this:

```dart
    (MyApp)
    ...
    onMapCreated((MapController mapController) {
      // We can now render the heatmap!
      mapController.upateHeatmap(heatmap);
    });
    ...
    Widget build(BuildContext ctx) {
       GoogleMap(
         onMapCreated: onMapCreated,
         // Potentially millions of other parameters (Circles, Polylines, Polygons...)
       )
    };
```

If the above doesn't add much extra latency to the rendering of the heatmap, we could be adding heatmaps in a backwards-compatible way. (I think)

If you want to keep the current API of passing the heatmap through the `GoogleMap` widget, I think the Platform Interface needs a little bit more tweaking, because the API that used to be _stringly_ typed is now **strOngly** typed :P

Thanks again for putting all the work to get this landed @fergusdixon!

**Comment by fergusdixon** [comment]
Sorry for the wait, busy week!

Yea definitely happy to make it work via the controller only, I think it's a bit more flexible than manipulating a variable in a `setState`.
From what I can see it would not change anything for the platform interface (this PR), it will just slightly change where the updates are triggered from, and how the end-user interacts with it I think?
I would also like to avoid adjusting @SvenSlijkoord's PR, and rather leave that work to the next PR which would be implementing the controller methods as well as 1 of the implementations (probably android?)

If we go the way of controller only, please let me know if you think I need to make any more changes to the current work to land it?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart' show listEquals;
+import 'package:flutter/material.dart' show Color;
+import 'package:meta/meta.dart' show immutable;
```
Comment (fergusdixon) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/heatmap.dart:11]:
```suggestion
import 'package:flutter/foundation.dart' show immutable;
```

**Comment by stuartmorgan** [comment]
@ditman It looks like this is ready for re-review?

**Comment by ditman** [comment]
Apologies, I was busy with some other stuff. Taking a look!

**Comment by ditman** [comment]
OK, my main concern with @SvenSlijkoord 's PR is that it _seems_ that he wanted to make Heatmaps work similarly to the rest of the geometry, see evidence here:

* https://github.com/flutter/plugins/pull/2454/files#diff-9bd2e43e18b051bf0a22c1fc3430348ca4b37755cdfd9b4dbb75a98ba156786bR225
* https://github.com/flutter/plugins/pull/2454/files#diff-9bd2e43e18b051bf0a22c1fc3430348ca4b37755cdfd9b4dbb75a98ba156786bR253

However, when Sven's published his PR, the `buildView` method looked like this:

* https://github.com/flutter/plugins/blob/10b40ddae933068914df0794953805d1a9283516/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_flutter_platform.dart#L340-L343

But now it looks like this:

* https://github.com/flutter/plugins/blob/9d32c83af9af125981ef847abd9d1636b65aa8e1/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_flutter_platform.dart#L352-L365

_Note how now, the `creationParams` (now `mapOptions`) do **not contain geometry** in them, and the initial geometry is passed through the buildView as named parameters._

_See also how the `serializeBlahSet` methods are [not used outside of the method channel implementation](https://github.com/flutter/plugins/search?q=serializeCircleSet) anymore, so they could be *not* exposed as new API from this change._

I think that before landing this, we need to see how the rest of Sven's PR will integrate with this package. My concern is that we might need to end up revisiting this code to add `heatmaps` to the `buildView` method of the platform interface (following the current approach), and that is a breaking change that would require a Major Version bump. If we can prove that the plugin will work with this code, we can land this as a non-breaking change.

/cc @stuartmorgan

**Comment by stuartmorgan** [comment]
That makes sense. Per https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins, we want to have a complete end-to-end implementation before we split out the package interface, to avoid these issues of landing something that potentially won't work. It sounds like we fundamentally don't have that here yet due to the current state of the main PR, in which case the best thing is to close this until we have an updated end-to-end PR (even if only for one platform).

@fergusdixon If you're interested in driving this forward, the next step would be to put together an updated end-to-end PR with at least one platform implementation, and then we can review that per the process described in the link above.

**Pull request closed without merging**